### PR TITLE
Feature/boas 690 publishing api

### PR DESCRIPTION
**Pull request by LetsMakeItTechnical**
## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [x] I have performed a self-review of my own code
- [x] I have double checked this work does not include any hardcoded secrets or passwords
- [ ] I have made corresponding changes to the documentation
- [ ] I have provided details on how I have tested my code
- [x] I have referenced the ticket number above
- [ ] I have provided a description of the ticket
- [x] I have included unit tests to cover any testable code changes

### CODE REVIEW

**Comment by francescobrenci-AR** [comment]
it should be merged to this branch `feature/boas-676-preview-and-publish-page`

File: apps/api/src/server/applications/application/__tests__/publish-application.test.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+test('publish an application/case', async (t) => {
+	const caseId = 1;
+
+	const response = await request.patch(`/applications/${caseId}/publish`);
+
+	t.is(response.status, 200);
+
+	t.deepEqual(response.body, {
+		publishedDate: mapDateStringToUnixTimestamp(new Date(1_649_319_144_000).toISOString())
```
Comment (chrisdalykinandcarta) [inline apps/api/src/server/applications/application/__tests__/publish-application.test.js:83]: I don't think we're testing the functionality of the method here, we're essentially just asserting that it was called. Would it be more useful to test against what we expect the string to be?

File: apps/api/src/server/applications/application/__tests__/publish-application.test.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	sinon.stub(Prisma.PrismaClient.prototype, '$transaction').returns([
+		{
+			publishedAt: new Date(1_649_319_144_000)
+		}
+	]);
+
+	sinon.useFakeTimers({ now: 1_649_319_144_000 });
+});
+
+test('publish an application/case', async (t) => {
```
Comment (mayya-bond-kc) [inline apps/api/src/server/applications/application/__tests__/publish-application.test.js:74]: I would add a few more test cases here. For example:
- Does an error get thrown if an invalid caseId is provided?
- Does the publishedAt date get updated if re-publishing?

File: apps/api/src/server/applications/applications.routes.js [hunk 3]
```diff
@@ -142,4 +143,33 @@ router.get(
...
+        }
+        #swagger.responses[200] = {
+            description: 'ID of application',
+            schema: { publishedDate }
+        }
+    */
+	validateApplicationId,
+	validateApplicantId,
+	validateCreateUpdateApplication,
+	trimUnexpectedRequestParameters,
```
Comment (mayya-bond-kc) [inline apps/api/src/server/applications/applications.routes.js:171]: Do we need these?

File: apps/api/src/server/repositories/case.repository.js [hunk 4]
```diff
@@ -232,6 +234,7 @@ const updateApplicationSansRegionsRemoval = ({
 	return databaseConnector.case.update({
 		where: { id: caseId },
 		data: {
+			publishedAt: publishedAt || null,
```
Comment (mayya-bond-kc) [inline apps/api/src/server/repositories/case.repository.js:237]: As we are adding a new piece for data to associate with a case, I don't think there is a matching migration for it

**Review by mayya-bond-kc — CHANGES_REQUESTED** [review]

File: apps/api/src/server/repositories/case.repository.js (hunk 4 above)
Comment (DavidJWhiteDev) [reply apps/api/src/server/repositories/case.repository.js:237]: I was wondering if that was the case as well, Mayya

File: apps/api/src/server/repositories/case.repository.js [hunk 5]
```diff
@@ -337,6 +340,38 @@ export const updateApplication = ({
...
+		where: { id: caseId },
+		data: {
+			publishedAt: new Date(),
+			modifiedAt: new Date()
+		}
+	});
+
+	transactions.push(application);
+
+	return databaseConnector.$transaction(transactions);
```
Comment (mayya-bond-kc) [inline apps/api/src/server/repositories/case.repository.js:359]: We don't actually need to use transactions here. Transactions are useful if we have multiple database operations that we would like to process, but here we are only doing a single `update` on `case`

Comment (LetsMakeItTechnical) [reply apps/api/src/server/repositories/case.repository.js:359]: ok will change this now

File: apps/api/src/server/applications/applications.routes.js [hunk 6]
```diff
@@ -142,4 +143,25 @@ router.get(
...
+        #swagger.description = 'publish application'
+        #swagger.parameters['id'] = {
+            in: 'path',
+			description: 'Application ID',
+			required: true,
+			type: 'integer'
+		}
+        #swagger.responses[200] = {
+            description: 'response will have the date that the case was published',
+            schema: { publishedDate }
```
Comment (DavidJWhiteDev) [inline apps/api/src/server/applications/applications.routes.js:160]: this does not display any response info in the swagger api, I would suggest:

#swagger.responses[200] = {
            description: 'response will have the date that the case was published as a timestamp',
            schema: { "publishedDate": 1673873105 }
        }

**Review by lutamKC — APPROVED** [review]

**Review by mayya-bond-kc — APPROVED** [review]
LGTM

**Pull request merged**